source/util/move_to_front.h

```diff
@@ -0,0 +1,318 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_UTIL_MOVE_TO_FRONT_H_
+#define LIBSPIRV_UTIL_MOVE_TO_FRONT_H_
+
+#include <cassert>
+#include <cstdint>
+#include <ostream>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace spvutils {
+
+// Log(n) move-to-front implementation. Implements two main functions:
+// IdFromRank - access id by its 0-indexed rank in the move-to-front sequence.
+// RankFromId - get the rank of the given id in the move-to-front sequence.
+// Accessing an id with any of the two functions moves the id to the front
+// of the sequence (rank of 0).
+//
+// The implementation is based on an AVL-based order statistic tree.
+//
+// Terminology
+// id: SPIR-V id.
+// rank: 0-indexed value showing how recently the id was accessed.
+// node: uint16_t handle used internally to access nodes.
+// size: size of the subtree of a node (including the node).
+// height: distance from a node to the farthest leaf.
+// index: not used to avoid confusion with rank and handle.
+class MoveToFront {
+  struct Node;
+ public:
+  explicit MoveToFront(size_t reserve_capacity = 128) {
+    nodes_.reserve(reserve_capacity);
+
+    // Create NIL node.
+    nodes_.emplace_back(Node());
+  }
+
+  // Returns the 0-indexed rank of id in the move-to-front sequence and moves
+  // id to the front. Example:
+  // Before the call: 4 8 2 1 7
+  // RankFromId(8) returns 1
+  // After the call: 8 4 2 1 7
+  size_t RankFromId(uint32_t id);
+
+  // Returns id corresponding to a 0-indexed rank in the move-to-front sequence
+  // and moves id to the front. Example:
+  // Before the call: 4 8 2 1 7
+  // IdFromRank(1) returns 8
+  // After the call: 8 4 2 1 7
+  uint32_t IdFromRank(size_t rank);
+
+  // Permanently removes the id from the move-to-front sequence.
```

Comment:
The method also generates an Id if rank is given as the current size.  (Is that a case we care about?  It seems weird to have this.)